numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -1149,85 +1359,125 @@ string_to_pycomplex(char *in, int has_null,
-            if (error_converting(complex_value.real)) {                          \
-                goto fail;                                                       \
-            }                                                                    \
-                                                                                 \
-            npy_csetreal##suffix(out, (npy_##ftype) complex_value.real);         \
-            npy_csetimag##suffix(out, (npy_##ftype) complex_value.imag);         \
-            in += in_stride;                                                     \
-            out += out_stride;                                                   \
-        }                                                                        \
-                                                                                 \
-        NpyString_release_allocator(allocator);                                  \
-        return 0;                                                                \
-                                                                                 \
-fail:                                                                            \
-        NpyString_release_allocator(allocator);                                  \
-        return -1;                                                               \
-        }                                                                        \
-                                                                                 \
-        static PyType_Slot s2##ctype##_slots[] = {                               \
-                {NPY_METH_resolve_descriptors,                                   \
-                 &string_to_##ctype##_resolve_descriptors},                      \
-                {NPY_METH_strided_loop, &string_to_##ctype},                     \
-                {0, NULL}};                                                      \
-                                                                                 \
-        static char *s2##ctype##_name = "cast_StringDType_to_" #ctype;
-
-STRING_TO_FLOAT_RESOLVE_DESCRIPTORS(cfloat, CFLOAT)
-STRING_TO_CFLOAT_CAST(cfloat, f, float)
-
-// cfloat to string
-
-FLOAT_TO_STRING_CAST(cfloat, cfloat, npy_cfloat)
-
-// string to cdouble
-
-STRING_TO_FLOAT_RESOLVE_DESCRIPTORS(cdouble, CDOUBLE)
-STRING_TO_CFLOAT_CAST(cdouble, , double)
-
-// cdouble to string
-
-FLOAT_TO_STRING_CAST(cdouble, cdouble, npy_cdouble)
-
-// string to clongdouble
-
-STRING_TO_FLOAT_RESOLVE_DESCRIPTORS(clongdouble, CLONGDOUBLE)
-STRING_TO_CFLOAT_CAST(clongdouble, l, longdouble)
-
-// longdouble to string
-
-FLOAT_TO_STRING_CAST(clongdouble, clongdouble, npy_clongdouble)
+template <
+    typename TNpyComplexType,
+    typename TNpyFloatType,
+    void npy_csetrealfunc(TNpyComplexType*, TNpyFloatType),
+    void npy_csetimagfunc(TNpyComplexType*, TNpyFloatType)
+>
+static int
+string_to_ctype(
```

Comment:
If you're changing the name, maybe `complex_float`? Both cfloat are and ctype are ambiguous if they're referring to the C language or complex numbers.